notification/tasks.py

```diff
@@ -0,0 +1,196 @@
+            )
+        except Exception as e:
+            logger.error(
+                f"Failed: An error occurred while sending Email notification error report for data-stewards."
+                f" Error: {e}"
+            )
+            print(
+                f"Failed: An error occurred while sending Email notification error report for data-stewards."
+                f" Error: {e}"
+            )
+
+
+@shared_task
+def send_notifications_for_user_upcoming_events(execution_date: str = None):
+    """
+    Send upcoming events notification report for all users, if any.
+
+    Params:
+        execution_date: The date of the execution of the task. FORMAT: YYYY-MM-DD (DEFAULT: Today)
+    """
+
+    logger.info("Sending notification for user upcoming events")
+
+    if not execution_date:
+        exec_date = datetime.now().date()
+    else:
+        exec_date = datetime.strptime(execution_date, "%Y-%m-%d").date()
+
+    users = get_user_model().objects.all()
+
+    for user in users:
+        notifications_exec_date = Notification.objects.filter(
+            recipient=user.id,
+            dispatch_by_email=True,
+            processing_date=None,
+            time__date=exec_date,
+        )
+
+        # send notification report to user, if any
+        if not notifications_exec_date:
+            # Checks if there is missed notification for this user and report errors to data-stewards, if any
+            report_notifications_upcoming_events_errors_for_datasteward(user)
+            continue
+
+        # group notifications per content type and set processed to today
+        notifications_by_content_type = defaultdict(list)
+        for notif in notifications_exec_date:
+            notifications_by_content_type[notif.content_type].append(notif)
+
+        context = {
+            "user": user.full_name,
+            "notifications": dict(notifications_by_content_type),
+        }
+        try:
+            send_the_email(
+                settings.EMAIL_DONOTREPLY,
+                user.email,
+                "Notifications",
+                "notification/email_list_notifications",
+                context,
+            )
+            for notif in notifications_exec_date:
+                notif.processing_date = datetime.now().date()
+                notif.save()
+        except Exception as e:
+            logger.error(
+                f"Failed: An error occurred while sending upcoming events Email notification for user {user.full_name}."
+                f" Error: {e}"
+            )
+            print(
+                f"Failed: An error occurred while sending upcoming events Email notification for user {user.full_name}."
+                f" Error: {e}"
+            )
+        finally:
+            # report error to data-stewards
+            report_notifications_upcoming_events_errors_for_datasteward(user)
+            continue
+
+
+@shared_task
+def send_all_notifications_for_user_upcoming_events(execution_date: str = None):
```

Comment:
I feel this is the exact same task as above, except for the filter applied to the Notification queryset. 

Would it make more sense to add a boolean parameter to the task (something like `one_day_only` which defaults to True for example)? When the parameter is true we set the filter `time__date=exec_date`, otherwise we set the filter `time__date__lte=exec_date`

Reply:
yes i had the same thought. I will do so. Thank you.